pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/SegmentMergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,199 @@
+
+      // Fetch the segment lineage for the table and compute the segments that should not be scheduled for merge
+      // based on the segment lineage.
+      SegmentLineage segmentLineageForTable = _clusterInfoProvider.getSegmentLineage(offlineTableName);
+      Set<String> segmentsNotToMerge = new HashSet<>();
+      if (segmentLineageForTable != null) {
+        for (String segmentLineageEntryId : segmentLineageForTable.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineageForTable.getLineageEntry(segmentLineageEntryId);
+          // Segments shows up on "segmentFrom" field in the lineage entry should not be scheduled again.
+          segmentsNotToMerge.addAll(lineageEntry.getSegmentsFrom());
+
+          // Segments shows up on "segmentsTo" field in the lineage entry with "IN_PROGRESS" state cannot be merged.
+          if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS) {
+            segmentsNotToMerge.addAll(lineageEntry.getSegmentsTo());
+          }
+        }
+      }
+
+      // Filter out the segments that cannot be merged
+      List<SegmentZKMetadata> segmentsToMergeForTable = new ArrayList<>();
+      List<String> scheduledSegments = scheduledSegmentsMap.getOrDefault(offlineTableName, Collections.emptyList());
+      for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : segmentsForOfflineTable) {
+        String segmentName = offlineSegmentZKMetadata.getSegmentName();
+
+        // The segment should not be merged if it's already scheduled or in progress
+        if (scheduledSegments.contains(segmentName) || segmentsNotToMerge.contains(segmentName)) {
+          continue;
+        }
+
+        // The segments that are newer than the buffer time period should not be be merged
+        if (System.currentTimeMillis() - offlineSegmentZKMetadata.getEndTime() < bufferTimePeriodMs) {
+          continue;
+        }
+        segmentsToMergeForTable.add(offlineSegmentZKMetadata);
+      }
+
+      // Compute Merge Strategy
+      List<List<SegmentZKMetadata>> segmentsToSchedule = MergeStrategyFactory.getMergeStrategy(taskConfigs)
+          .generateMergeTaskCandidates(segmentsToMergeForTable, maxNumSegmentsPerTask);
+
+      // Generate tasks
+      for (List<SegmentZKMetadata> segments : segmentsToSchedule) {
+        if (tableNumTasks == tableMaxNumTasks) {
+          break;
+        }
+        if (segments.size() >= 1) {
+          Map<String, String> configs = new HashMap<>();
+          configs.put(MinionConstants.TABLE_NAME_KEY, offlineTableName);
+          configs.put(MinionConstants.SEGMENT_NAME_KEY,
+              segments.stream().map(s -> s.getSegmentName()).collect(Collectors.joining(",")));
+          configs.put(MinionConstants.DOWNLOAD_URL_KEY,
+              segments.stream().map(s -> ((OfflineSegmentZKMetadata) s).getDownloadUrl())
+                  .collect(Collectors.joining(",")));
+          configs.put(MinionConstants.VIP_URL_KEY, _clusterInfoProvider.getVipUrl());
```

Comment:
Should this be `uploadURL`?

Reply:
I get the `VIP_URL` instead of `uploadURL` because I also need to compute the URL for `startReplaceSegment` and `endReplaceSegment`

So, I get the vip url, which is the base url, and I compute URLs for all APIs that I'm calling.